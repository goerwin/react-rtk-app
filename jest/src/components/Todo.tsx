import { ChangeEvent, FormEvent, useRef, useState } from 'react';
import {
  useAddNewTodoMutation,
  useEditTodoMutation,
  useGetTodosQuery,
  useRemoveTodoMutation,
} from '../features/api/apiSlice';
import LoadingSpinner from './LoadingSpinner';
import styles from './Todo.module.scss';

export default function Todo() {
  const inputElRef = useRef<HTMLInputElement>(null);
  const [inputElValue, setInputElValue] = useState('');
  const { isFetching: isTodosLoading, data: todos, isError, error } = useGetTodosQuery();
  const [addNewTodo, { isLoading: isAddNewTodoLoading }] = useAddNewTodoMutation();
  const [removeTodo, { isLoading: isRemoveTodoLoading }] = useRemoveTodoMutation();
  const [editTodo, { isLoading: isEditTodoLoading }] = useEditTodoMutation();

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputElValue(e.target.value);
  };

  const handleFormSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const value = inputElRef.current?.value;
    if (!value || isAddNewTodoLoading) return;

    await addNewTodo(value);
    setInputElValue('');
  };

  const handleEditTodo = async (todo: { id: string; completed: boolean }) => {
    await editTodo({ id: todo.id, todo: { ...todo, completed: !todo.completed } });
  };

  const handleRemoveTodo = async (id: string) => {
    await removeTodo(id);
  };

  if (isError) {
    return (
      <main className={styles.container}>
        <p data-testid="error-message">{(error as any).data || 'Error Loading Todos'}</p>
      </main>
    );
  }

  return (
    <main className={styles.container}>
      <LoadingSpinner
        show={isTodosLoading || isAddNewTodoLoading || isRemoveTodoLoading || isEditTodoLoading}
      />

      <div className={styles.formContainer}>
        <h1>Todo (Asynchronous actions)</h1>

        <form onSubmit={handleFormSubmit}>
          <input
            ref={inputElRef}
            value={inputElValue}
            onChange={handleInputChange}
            placeholder="Add Todo"
          />
          <button>Add</button>
        </form>
      </div>

      <ul className={styles.todosContainer}>
        {todos?.map((el) => (
          <li key={el.id}>
            <p className={`${el.completed ? styles.isCompleted : ''}`}>{el.title}</p>

            <button
              onClick={() => handleEditTodo(el)}
              className={`${el.completed ? styles.isSecondary : ''}`}
            >
              {el.completed ? 'Not Done' : 'Done'}
            </button>

            <button onClick={() => handleRemoveTodo(el.id)} className={styles.isDanger}>
              Remove
            </button>
          </li>
        ))}
      </ul>
    </main>
  );
}
