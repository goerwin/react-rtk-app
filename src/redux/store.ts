import { configureStore } from '@reduxjs/toolkit';
import i18nReducer from '../features/i18n/i18nSlice';

export const store = configureStore({
  reducer: {
    i18n: i18nReducer,
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
