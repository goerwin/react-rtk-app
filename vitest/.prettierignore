coverage
jest/coverage
vitest/coverage
dist
node_modules
