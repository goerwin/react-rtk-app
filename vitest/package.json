{
  "name": "react-rtk-app",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "build": "tsc && vite build",
    "dev": "vite --host",
    "preview": "vite preview",
    "test": "vitest run",
    "test-watch": "vitest",
    "test-coverage": "vitest run --coverage",
    "prettier-check": "prettier --check .",
    "prettier-write": "prettier --write ."
  },
  "dependencies": {
    "@reduxjs/toolkit": "^1.9.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-redux": "^8.0.4",
    "whatwg-fetch": "^3.6.2"
  },
  "devDependencies": {
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/react": "^18.0.22",
    "@types/react-dom": "^18.0.7",
    "@vitejs/plugin-react": "^2.2.0",
    "@vitest/coverage-c8": "^0.24.4",
    "autoprefixer": "^10.4.13",
    "jsdom": "^20.0.2",
    "msw": "^0.48.3",
    "prettier": "^2.7.1",
    "sass": "^1.56.1",
    "typescript": "^4.6.4",
    "vite": "^3.2.4",
    "vitest": "^0.24.5"
  },
  "msw": {
    "workerDirectory": "public"
  }
}
