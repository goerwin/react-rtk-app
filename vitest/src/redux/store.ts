import { configureStore } from '@reduxjs/toolkit';
import counterReducer from '../features/counter/counterSlice';
import { apiSlice } from '../features/api/apiSlice';

export function createStore() {
  return configureStore({
    reducer: {
      counter: counterReducer,
      [apiSlice.reducerPath]: apiSlice.reducer,
    },

    // rtk-query needs to be added as last middleware
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiSlice.middleware),
  });
}

export const store = createStore();

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
