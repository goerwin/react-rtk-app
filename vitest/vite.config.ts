/// <reference types="vitest" />

import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig({
  plugins: [react()],
  test: {
    environment: 'jsdom',
    include: ['**/*.{test,spec}.{ts,tsx}'],
    setupFiles: ['tests/index.ts'],
    coverage: {
      provider: 'c8',
      include: ['src/**/*.{ts,tsx}'],
      exclude: ['src/mocks/**/*', 'src/**/*.test.{ts,tsx}', '**/*.d.ts'],

      // include all files, including the untested ones into report.
      all: true,

      // minimum threshold to meet
      statements: 0,
      functions: 0,
      branches: 0,
      lines: 90,
    },
  },
});
