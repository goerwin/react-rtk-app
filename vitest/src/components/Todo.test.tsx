import { render, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { Provider } from 'react-redux';
import { expect, it } from 'vitest';
import { mockServer } from '../../tests/mockServer';
import { createStore } from '../redux/store';
import Todo from './Todo';

it('should show a loading state', () => {
  const screen = render(
    <Provider store={createStore()}>
      <Todo />
    </Provider>
  );

  expect(screen.asFragment()).toMatchSnapshot();
});

it('should load 3 todo items after loading is done', async () => {
  const screen = render(
    <Provider store={createStore()}>
      <Todo />
    </Provider>
  );

  // wait for loading to go away
  await waitFor(() => expect(screen.queryByTestId('global-loading')).toBeNull());

  // get all listitem (li) elements in the page
  const list = screen.getByRole('list');
  const items = within(list).getAllByRole('listitem');

  expect(items.length).toEqual(3);
});

it('should show error message on API error', async () => {
  const screen = render(
    <Provider store={createStore()}>
      <Todo />
    </Provider>
  );

  // mock get /todos with a 500 error
  mockServer.use(
    rest.get('https://jsonplaceholder.typicode.com/todos', async (req, res, ctx) => {
      return res(ctx.status(500), ctx.text('Error loading todos'));
    })
  );

  // wait for loading to go away
  await waitFor(() => expect(screen.queryByTestId('global-loading')).toBeNull());

  // verify error message is displayed
  expect(screen.getByTestId('error-message').textContent).toEqual('Error loading todos');
});

// Test skipped to show coverage
it.skip('should remove the second item after clicking remove', async () => {
  const screen = render(
    <Provider store={createStore()}>
      <Todo />
    </Provider>
  );

  // wait for loading to go away
  await waitFor(() => expect(screen.queryByTestId('global-loading')).toBeNull());

  // get all listitem (li) elements in the page
  const list = screen.getByRole('list');
  const items = within(list).getAllByRole('listitem');

  // verify we still have 3 todo items
  expect(items.length).toEqual(3);

  // get remove button of 2nd item and click it
  const removeBtn = within(items[1]).getByRole('button', { name: 'Remove' });
  await userEvent.click(removeBtn);

  // wait for loading to go away
  await waitFor(() => expect(screen.queryByTestId('global-loading')).toBeNull());

  // verify there are now 2 items
  const newItems = within(list).getAllByRole('listitem');
  expect(newItems.length).toEqual(2);
});
