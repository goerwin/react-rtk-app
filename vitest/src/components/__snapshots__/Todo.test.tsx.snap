// Vitest Snapshot v1

exports[`should show a loading state 1`] = `
<DocumentFragment>
  <main
    class="_container_652d7f"
  >
    <div
      aria-busy="true"
      class="_container_2e4c81"
      data-testid="global-loading"
      role="alert"
    >
      <div
        class="_spinner_2e4c81"
      />
    </div>
    <div
      class="_formContainer_652d7f"
    >
      <h1>
        Todo (Asynchronous actions)
      </h1>
      <form>
        <input
          placeholder="Add Todo"
          value=""
        />
        <button>
          Add
        </button>
      </form>
    </div>
    <ul
      class="_todosContainer_652d7f"
    />
  </main>
</DocumentFragment>
`;
